proposals/NNNN-isolated-synchronous-deinit.md

```diff
@@ -0,0 +1,386 @@
+} 
+```
+
+In case of escaping self, race condition is eliminated but problem of escaping `self` remains. This problem exists for synchronous code as well and is orthogonal to the concurrency features.
+
+```swift
+actor Clicker {
+  var count: Int = 0
+
+  func click(_ times: Int) {
+    for _ in 0..<times {
+      self.count += 1 
+    }
+  }
+
+  deinit {
+    let old = count
+    let moreClicks = 10000
+    
+    Task { await self.click(moreClicks) } // ❌ This WILL keep `self` alive after the `deinit`!
+
+    for _ in 0..<moreClicks {
+        // No data race.
+        // Actor job created by the task is either
+        // not created yet or is waiting in the queue.
+      self.count += 1 
+    }
+
+    assert(count == old + moreClicks) // Always works
+    assert(count == old + 2 * moreClicks) // Always fails
+  }
+}
+```
+
+## Detailed design
+
+### Runtime
+
+Proposal introduces new runtime function that is used to schedule task-less block of synchronous code on the executor by wrapping it into an ad hoc task. It does not do any reference counting and can be safely used even with references that were released for the last time but not deallocated yet.
+
+If no switching is needed, block is executed immediately on the current thread. Otherwise, task-less job copies priority and task-local values from the task/thread that released the last strong reference to the object.
+
+```cpp
+using AdHocWorkFunction = SWIFT_CC(swift) void (void *);
+  
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+void swift_task_performOnExecutor(void *context, AdHocWorkFunction *work, ExecutorRef newExecutor);
+```
+
+```swift
+@_silgen_name("swift_task_performOnExecutor")
+@usableFromInline
+internal func _performOnExecutor(_ ctx: __owned AnyObject,
+                               _ work: @convention(thin) (__owned AnyObject) -> Void,
+                               _ executor: UnownedSerialExecutor)
+```
+
+If `deinit` is isolated, code that normally is emitted into `__deallocating_init` gets emitted into new entity - `__isolated_deallocating_init`. And `__deallocating_init` is emitted as a thunk that reads executor from `self` (for actors) or global actor (for GAITs) and calls `swift_task_performOnExecutor` passing `self`, `__isolated_deallocating_init` and desired executor.
+
+![Last release from background thread](https://aws1.discourse-cdn.com/swift/original/3X/a/0/a06988cfffdacaef2387da035789f965c2345c5b.png)
+
+![Dealloc on main thread](https://aws1.discourse-cdn.com/swift/original/3X/d/8/d8098cdc104b2c87072d63a22060106c4dc5b4e7.png)
+
+Non-deallocting deinit is not affected by the changes.
+
+### Rules for computing isolation
+
+Isolation of deinit comes with runtime and code size cost. Classes that don't perform custom actions in deinit and only need to release references don't need isolated deinit. Releasing child objects can be done from any thread. If those objects are concerned about isolation, they should adopt isolation themselves.
```

Comment:
> "don't need isolated deinit" really does mean "will not receive isolated deinit under this proposal."

Not exactly. My intention was to explain rationale for excluding implicit `deinit` from isolation, rather than just stating the decision.

> And does "don't perform custom actions" just refer to explicit vs. implicit deinit?

Yes

Reply:
Okay, that makes sense to me. I think this is fine to leave as-is on that front, though I think the suggested change in the first comment here is still relevant.